lib/puppet/util/windows/registry.rb

```diff
@@ -31,16 +31,24 @@ def open(name, path, mode = KEY_READ | KEY64, &block)
     end
 
     def values(subkey)
+      orig_enc = Encoding.default_internal
+      Encoding.default_internal = Encoding::UTF_8
```

Comment:
So the problem here @joshcooper is that inside of Ruby's registry code, they're doing some puzzling stuff.

In their `each_key` and `each_value` methods they're using the [RegEnumKeyEx](http://msdn.microsoft.com/en-us/library/windows/desktop/ms724862\(v=vs.85\).aspx) and [RegEnumValueW](http://msdn.microsoft.com/en-us/library/windows/desktop/ms724865\(v=vs.85\).aspx) API calls respectively.

As expected, the values returned from these calls include some `lpName [out]` which are `UTF-16LE`.  So far, so good.  Verified during 2.1.5 debugging, the string has the appropriate encoding after the API call returns.

Unfortunately, https://github.com/ruby/ruby/commit/79e880c90df6938e8226336a883d1973031da3df added the `export_string` function.  (For reference, see the other "interesting" commits following that one via https://github.com/ruby/ruby/commits/802d4f9fb2a8a5f2be1eb6be13dc4f286ad221af/ext/win32/lib/win32/registry.rb):

```ruby
def export_string(str, enc = Encoding.default_internal || LOCALE) # :nodoc:
  str.encode(enc)
end
```

As @ferventcoder pointed out, `Encoding.default_internal` will typically be `nil` by default, and Ruby hardcodes `LOCALE` with `LOCALE = Encoding.find(Encoding.locale_charmap)`.  The `LOCALE` here is the fallback, and is set based on current code page - which will be system dependent.  On my system, default is `1252` and everything is OK, but on other systems it could be `437` which means certain `UTF-16LE` characters in reg key or value names (like endash) might not translate properly.

From `each_value`, we can see that we have a `UTF-16LE` conversion to `LOCALE`:

```ruby
begin
  #### Wrapper for RegEnumValueEx that returns UTF-16LE
  subkey = API.EnumValue(@hkey, index)
rescue Error
  break
end
##### TRANSLATE out of UTF-16LE to LOCALE
subkey = export_string(subkey)
begin
  type, data = read(subkey)
rescue Error
  next
end
```

And then of course, when we look at what [read](https://github.com/ruby/ruby/blob/802d4f9fb2a8a5f2be1eb6be13dc4f286ad221af/ext/win32/lib/win32/registry.rb#L641) and [QueryValue](https://github.com/ruby/ruby/blob/802d4f9fb2a8a5f2be1eb6be13dc4f286ad221af/ext/win32/lib/win32/registry.rb#L315-L323) (wrapper over [RegQueryValueEx](http://msdn.microsoft.com/en-us/library/windows/desktop/ms724911%28v=vs.85%29.aspx)) do, it's more evident why this is problematic.  As you might expect, `make_wstr` calls `str.encode(WCHAR)` where `WCHAR = Encoding::UTF_16LE`.

```ruby
def read(name, *rtype)
  #### now call RegQueryValueEx with a LOCALE encoded string
  type, data = API.QueryValue(@hkey, name)
  unless rtype.empty? or rtype.include?(type)
    raise TypeError, "Type mismatch (expect #{rtype.inspect} but #{type} present)"
  end
  case type
  when REG_SZ, REG_EXPAND_SZ
    #### encode accepts dst_encoding, src_encoding
    #### so this translates UTF-16LE data back to LOCALE
    [ type, data.encode(name.encoding, WCHAR).chop ]
  when REG_MULTI_SZ
    [ type, data.encode(name.encoding, WCHAR).split(/\0/) ]

#### SNIP

def QueryValue(hkey, name)
  type = packdw(0)
  size = packdw(0)
  #### LOCALE encoded string turned back into UTF-16LE
  name = make_wstr(name)
  check RegQueryValueExW.call(hkey, name, 0, type, 0, size)
  data = "\0".force_encoding('ASCII-8BIT') * unpackdw(size)
  check RegQueryValueExW.call(hkey, name, 0, type, data, size)
  [ unpackdw(type), data[0, unpackdw(size)] ]
end
```

As you can see, Ruby has:

* Read the value names via `RegEnumValueW` as `UTF-16LE`
* Translated those to `LOCALE`, whatever that may be
* Round-tripped back to `UTF-16LE` prior to calling `RegQueryValueExW`
* Depending on the type of the `data` returned, a `UTF-16LE` string value is translated back to `LOCALE`

Clearly the right thing to do here is to leave the `UTF-16LE` string alone prior to calling `RegQueryValueExW`, and it looks like the only good way to leave that string intact and not introduce unnecessary roundtrips is:

- Set `Encoding.default_internal` to `UTF-16LE`
- Set `LOCALE` to `UTF-16LE`
- Monkey patch the Ruby code so that `export_string` isn't re-encoding strings unnecessarily as this appears to be the only affected call-site

These can be done at a global level, which seems rather dangerous, or they can be done at a more local level when calling into these Ruby methods.  The safer approach is clearly (as long as there's a proper try/catch/finally style mechanism) to do this in a more local level.  It's possible Ruby will realize the error of this approach and fix in a later version, so patching `export_string` seems dangerous as well.

Also note, that there is another spot where `LOCALE` is forced that could be problematic (and requires additional investigation), [self.expand_environ](https://github.com/ruby/ruby/blob/802d4f9fb2a8a5f2be1eb6be13dc4f286ad221af/ext/win32/lib/win32/registry.rb#L372-L378) which is used whenever `REG_EXPAND_SZ` values are encountered.

```ruby
def self.expand_environ(str)
  str.gsub(Regexp.compile("%([^%]+)%".encode(str.encoding))) {
    v = $1.encode(LOCALE)
    (e = ENV[v] || ENV[v.upcase]; e.encode(str.encoding) if e) ||
    $&
  }
end
```

Sigh, Ruby.